catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright (c) Codice Foundation
+ * <p/>
+ * This is free software: you can redistribute it and/or modify it under the terms of the GNU Lesser
+ * General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or any later version.
+ * <p/>
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.util.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingDeque;
+
+import org.apache.commons.lang3.Validate;
+
+import ddf.catalog.CatalogFramework;
+import ddf.catalog.data.Result;
+import ddf.catalog.federation.FederationException;
+import ddf.catalog.operation.impl.QueryImpl;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+
+public class QueryResultPaginator {
+
+    private final CatalogFramework catalogFramework;
+
+    private final QueryImpl queryCopy;
+
+    private int indexOffset;
+
+    private int pageSize;
+
+    //Set to true when a query returns zero results.
+    private boolean isExhausted;
+
+    private BlockingQueue<Result> queriedResultsBuffer;
+
+    /**
+     * QueryResultPaginator takes in the CatalogFramework object and a Query Request and
+     * queries for the number of results needed. This is done using the pageSize specified
+     * in the Query object. The Paginator will return a collection of results of the
+     * specified page size.
+     * <p>
+     * The startIndex is the first index in the results gathered from the CatalogFramework
+     * <p>
+     * The endIndex is the last index in the results gathered from the CatalogFramework, which
+     * may not be reached if the pageSize is smaller than the difference between the endIndex
+     * and startIndex.
+     * <p>
+     * The resultsRemaining are the total results not yet returned in a collection of the size
+     * of a page. This value is tracked through the hasNextPage() method to inform the client that
+     * there are valid results to continue querying.
+     *
+     * @param catalogFramework
```

Comment:
❗️ We should add descriptions here. And we may want to put the `startIndex` and `pageSize` comments on the `query` parameter description instead of in the class description.